### PR TITLE
[4.0] Replace the zero datetime value '0000-00-00 00:00:00' by the real NULL value

### PR DESCRIPTION
**Pull request by csthomas**
Pull Request for Issue # .

### Summary of Changes
This PR changes only datetimes in the module table. 
No more zero dates like "0000-00-00 00:00:00" in the module table.

Method `checkIn()` is only added (almost copy/paste) to set column `check_out_time` to NULL.
At method `store()` the default value has been changed to `$updateNulls = true` to be able to save `NULL` values.

### TODO
- ~~missing changes for the install files (Mysql, PostgreSQL)~~
- ~~missing changes in the update files for Postgresql version~~

Changes for other tables should be made in the next PRs

### Testing Instructions
**First way:**

1. On existed 4.0-dev installation upload changes and then click on the database "Fix" button.
    The `UPDATE ...` query won't be applied by the `FIX` button. You can do it manually (in phpMyAdmin), but it is not required, the system will work anyway!
    - on Postgresql database Fix button does not work because PR https://github.com/joomla/joomla-cms/pull/19707 has not yet been merged  in 4.0.

**Second way:**

1. Upload changes, then remove file `configuration.php` and install joomla 4.0 again (also you can test sample data).

**In both cases:**

2. After changes will be applied you can create and save modules with real NULL dates. 
    Test publish_up, `publish_down` and `check_out_time` columns. Everything work as before.

### Expected result
You can create, edit and save modules. 
Modules are loaded normally on backend and frontend.

### How to revert changes in your test db
```sql
SET sql_mode = '';
ALTER TABLE `#__modules` MODIFY `publish_up` DATETIME NOT NULL DEFAULT '0000-00-00 0000:00:00';
ALTER TABLE `#__modules` MODIFY `publish_down` DATETIME NOT NULL DEFAULT '0000-00-00 0000:00:00';
ALTER TABLE `#__modules` MODIFY `checked_out_time` DATETIME NOT NULL DEFAULT '0000-00-00 0000:00:00';
```

### CODE REVIEW

**Comment by laoneo** [comment]
I like the change. I was always wondering why we bloat the rows with such 0 dates instead of null.

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-08-29.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+ALTER TABLE `#__modules` MODIFY `publish_up` datetime NULL DEFAULT NULL;
+ALTER TABLE `#__modules` MODIFY `publish_down` datetime NULL DEFAULT NULL;
+ALTER TABLE `#__modules` MODIFY `checked_out_time` datetime NULL DEFAULT NULL;
+
+-- Use 0 instead of '0000-00-00 00:00:00' to not get 'Invalid default value for ...'
+UPDATE `#__modules` SET
+	`publish_up` = CASE WHEN `publish_up` IN (0, '1000-01-01 00:00:00') THEN NULL ELSE `publish_up` END,
```
Comment (alikon) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-08-29.sql]: shouldn't be `IN (0, '1000-01-01 00:00:00', '0000-00-00 00:00:00')`

**Comment by alikon** [comment]
we should do this now in 4.0 or never more

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-08-29.sql (hunk 1 above)
Comment (csthomas) [reply administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-08-29.sql]: `0 == '0000-00-00 00:00:00'` in mysql. This is almost the same.

If we want to be strict then it should be `IN ('0000-00-00 00:00:00', '1000-01-01 00:00:00')` but then in phpmyadmin on mysql 5.7 I got error `Invalid default value for ...`

Comment (alikon) [reply administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-08-29.sql]: yes cause phpmyadmin have a standard/default  `sql_mode` for mysql > 5.7

but for mysql server < 5.7 is always true that `0 == '0000-00-00 00:00:00'`  ?
not sure depending on `sql_mode` settings

**Comment by csthomas** [comment]
Ready for core review and test:)

**Comment by csthomas** [comment]
By adding patch from `curl https://patch-diff.githubusercontent.com/raw/joomla/joomla-cms/pull/19707.diff | git apply --reject` you can test the update file for postgresql

File: libraries/src/Table/Module.php [hunk 2]
```diff
@@ -179,31 +191,106 @@ public function bind($array, $ignore = '')
...
+	 * @since   __DEPLY_VERSION__
+	 * @throws  \UnexpectedValueException
 	 */
-	public function store($updateNulls = false)
+	public function checkIn($pk = null)
 	{
-		// Set publish_up, publish_down and checked_out_time to null date if not set
-		if (!$this->publish_up)
+		// Pre-processing by observers
+		$event = AbstractEvent::create(
```
Comment (laoneo) [inline libraries/src/Table/Module.php]: Can we do the events in a new pr, so we can focus here on the null issue?

**Comment by csthomas** [comment]
In the "Files changed" tab it is probably not visible. I copy&paste `checkIn()` method from Table.php and pasted it into `Module.php`. Then I changed only the comment and two lines to change the sql zero date with `NULL` and empty string with `null`.

I will change the order of methods in the file, and the diff view should be more readable.

**Comment by csthomas** [comment]
The diff between Table.php and Module.php for `checkIn()` method. 

screenshot_20180831_113733

I did not want to add changes in Table.php because it breaks other tables.

**Comment by laoneo** [comment]
Would be good if we can do it without copy paste. Can you not expand the query with an "or is null" in Table.php?

**Comment by csthomas** [comment]
The other way may be to add some class value like `Table::supportNullDates = false` for all and `Module::supportNullDates = true`.

@mbabker Can you advise?

**Comment by csthomas** [comment]
> Can you not expand the query with an "or is null" in Table.php?

Read carefully, this is an update query (`SET ....`):
```diff
- ->set($this->_db->quoteName($checkedOutTimeField) . ' = ' . $this->_db->quote($this->_db->getNullDate()));
+ ->set($this->_db->quoteName($checkedOutTimeField) . ' = NULL');
```

**Comment by csthomas** [comment]
I think about something similar to:
```diff
diff --git a/libraries/src/Table/Table.php b/libraries/src/Table/Table.php
index a027835e8b..3828aeed34 100644
--- a/libraries/src/Table/Table.php
+++ b/libraries/src/Table/Table.php
@@ -122,6 +122,14 @@ abstract class Table extends CMSObject implements \JTableInterface, DispatcherAw
         */
        protected $_jsonEncode = array();
 
+       /**
+        * Indicate that columns full support database NULL value
+        *
+        * @var    boolean
+        * @since  __DEPLOY_VERSION__
+        */
+       protected $_supportNullValue = false;
+
        /**
         * The UCM type alias. Used for tags, content versioning etc. Leave blank to effectively disable these features.
         *
```

**Comment by csthomas** [comment]
I changed the code and now should be more readable, without overloading `checkIn()` method.

**Review by laoneo — APPROVED** [review]

**Comment by csthomas** [comment]
@alikon @tonypartridge @ggppdk @Bakual I am asking you for a test, at least for a part, I know people are very busy. 


If this go then I can work on the rest of tables.

**Comment by ggppdk** [comment]
I have tested this item :white_check_mark: successfully on b3e640cdb9104fa6abb7cd008c93b819f908aafd<br /><br />@csthomas thanks for making such PRs to improve / fix Database API and Database handling

I have tested with MySQL
frontend and backend modules,

also looked to see if there are more places to be updated with usage of `isNullDatetime()` by looking at files that have `#__modules as`
but did not find any more places <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21901">issues.joomla.org/tracker/joomla-cms/21901</a>.</sub>

**Comment by alikon** [comment]
on my to do list

**Comment by laoneo** [comment]
Can we get here some tests? Would be nice to get it in.

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on a692e026e7171f4326ee8a13a01375a010a45685<br /><br />sorry for the delay<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21901">issues.joomla.org/tracker/joomla-cms/21901</a>.</sub>

**Comment by csthomas** [comment]
There is some conflicts so I have to resolve it. @wilsonge do you have time to review the code?

**Comment by csthomas** [comment]
I rebased PR. There were conflicts in sql files. Now there is need one more test.

**Review by wilsonge — APPROVED** [review]

**Comment by laoneo** [comment]
Would be good to get some more tests here and the conflicts fixed, so we can get it merged.

**Comment by csthomas** [comment]
I did what I could. Your turn.

**Review by wilsonge — APPROVED** [review]

**Comment by ggppdk** [comment]
https://github.com/joomla/joomla-cms/issues/24151

So what is the status with this PR ?
Furthermore after this gets merged , more similar PRs will be needed

**Comment by wilsonge** [comment]
Let's go for this. I think we still have other compatibility issues. But this seems like a positive starting point.

**Comment by ghost** [comment]
@ggppdk can #24151 be closed or needs more Pull Requests?

**Comment by wilsonge** [comment]
@franz-wohlkoenig needs more. this rolls it out for modules. this will need rolling out everywhere that has a publish up/down column (com_content etc etc)

**Comment by hardik-codes** [comment]
Members may I help with further PRs ?
@wilsonge @franz-wohlkoenig

**Comment by wilsonge** [comment]
Fine by me

**Comment by hardik-codes** [comment]
> Fine by me

Where in are the further changes to be made?

**Comment by sakiss** [comment]
Tested that and works.
Fixes the error messages related with empty `published_up`, `published_down` and `checked_out_time` in the module saving as well.

**Pull request merged**